File: archive/2024-06-01/tree/T2T3T4_orderTraversal.java
package org.review2024summer.tree;

import java.util.*;

public class T2T3T4_orderTraversal {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        preorder(root, list);
        return list;
    }

    public void preorder(TreeNode root, List<Integer> list) {
        if (root == null)
            return;
        list.add(root.val);
        preorder(root.left, list);
        preorder(root.right, list);
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        inorder(root, list);
        return list;
    }

    public void inorder(TreeNode root, List<Integer> list) {
        if (root == null)
            return;
        inorder(root.left, list);
        list.add(root.val);
        inorder(root.right, list);
    }

    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        postorder(root, list);
        return list;
    }

    public void postorder(TreeNode root, List<Integer> list) {
        if (root == null)
            return;
        postorder(root.left, list);
        postorder(root.right, list);
        list.add(root.val);
    }

    public List<Integer> preorderTraversalByIter(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        if (root == null)
            return list;

        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode tmp = stack.pop();
            list.add(tmp.val);
            if (tmp.right != null)
                stack.push(tmp.right);
            if (tmp.left != null)
                stack.push(tmp.left);
        }
        return list;
    }

    public List<Integer> postorderTraversalByIter(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        if (root == null)
            return list;

        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode tmp = stack.pop();
            list.add(tmp.val);
            if (tmp.left != null)
                stack.push(tmp.left);
            if (tmp.right != null)
                stack.push(tmp.right);
        }
        Collections.reverse(list);
        return list;
    }

    public List<Integer> inorderTraversalByIter(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        if (root == null)
            return list;

        TreeNode curr = root;
        while (!(curr == null && stack.isEmpty())) {
            if (curr != null) {
                stack.push(curr);
                curr = curr.left;
            } else {
                curr = stack.pop();
                list.add(curr.val);
                curr = curr.right;
            }
        }

        return list;
    }

    public List<Integer> inorderTraversalByMonitor(TreeNode root) {
        ArrayDeque<TreeNode> stack = new ArrayDeque<>();
        ArrayList<Integer> list = new ArrayList<>();
        final int VISITED = Integer.MAX_VALUE;
        if (root == null)
            return list;

        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode curr = stack.pop();
            if (curr.val != VISITED) {
                if (curr.right != null) {stack.push(curr.right);}
                stack.push(curr);
                stack.push(new TreeNode(VISITED));
                if (curr.left != null) {stack.push(curr.left);}
            } else {
                list.add(stack.pop().val);
            }
        }
        return list;
    }

    public List<Integer> preorderTraversalByMonitor(TreeNode root) {
        ArrayDeque<TreeNode> stack = new ArrayDeque<>();
        ArrayList<Integer> list = new ArrayList<>();
        final int VISITED = Integer.MAX_VALUE;
        if (root == null)
            return list;

        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode curr = stack.pop();
            if (curr.val != VISITED) {
                if (curr.right != null) {stack.push(curr.right);}
                if (curr.left != null) {stack.push(curr.left);}
                stack.push(curr);
                stack.push(new TreeNode(VISITED));
            } else {
                list.add(stack.pop().val);
            }
        }
        return list;
    }

    public List<Integer> postorderTraversalByMonitor(TreeNode root) {
        ArrayDeque<TreeNode> stack = new ArrayDeque<>();
        ArrayList<Integer> list = new ArrayList<>();
        final int VISITED = Integer.MAX_VALUE;
        if (root == null)
            return list;

        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode curr = stack.pop();
            if (curr.val != VISITED) {
                stack.push(curr);
                stack.push(new TreeNode(VISITED));
                if (curr.right != null) {stack.push(curr.right);}
                if (curr.left != null) {stack.push(curr.left);}
            } else {
                list.add(stack.pop().val);
            }
        }
        return list;
    }





}

File: archive/2026-04-05/ch06_tree/t2_t3_t4_order_traversal.py
"""LeetCode Problems:
- 144. Binary Tree Preorder Traversal
  https://leetcode.cn/problems/binary-tree-preorder-traversal/
- 94. Binary Tree Inorder Traversal
  https://leetcode.cn/problems/binary-tree-inorder-traversal/
- 145. Binary Tree Postorder Traversal
  https://leetcode.cn/problems/binary-tree-postorder-traversal/
"""

from typing import List, Optional

from tree_node import TreeNode


class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        self.preorder(root, ans)
        return ans

    def preorder(self, root: Optional[TreeNode], ans: List[int]) -> None:
        if not root:
            return
        self.preorder(root.left, ans)
        ans.append(root.val)
        self.preorder(root.right, ans)


    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        pass

    def inorder(self, root: Optional[TreeNode], list: List[int]) -> None:
        pass

    def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        pass

    def postorder(self, root: Optional[TreeNode], list: List[int]) -> None:
        pass

    def preorderTraversalByIter(self, root: Optional[TreeNode]) -> List[int]:
        pass

    def postorderTraversalByIter(self, root: Optional[TreeNode]) -> List[int]:
        pass

    def inorderTraversalByIter(self, root: Optional[TreeNode]) -> List[int]:
        pass

    def inorderTraversalByMonitor(self, root: Optional[TreeNode]) -> List[int]:
        pass

    def preorderTraversalByMonitor(self, root: Optional[TreeNode]) -> List[int]:
        pass

    def postorderTraversalByMonitor(self, root: Optional[TreeNode]) -> List[int]:
        pass


if __name__ == "__main__":
    root = TreeNode(1, None, TreeNode(2, TreeNode(3)))
    solution = Solution()

    assert solution.preorderTraversal(root) == [1, 2, 3]
    assert solution.inorderTraversal(root) == [1, 3, 2]
    assert solution.postorderTraversal(root) == [3, 2, 1]

    assert solution.preorderTraversalByIter(root) == [1, 2, 3]
    assert solution.inorderTraversalByIter(root) == [1, 3, 2]
    assert solution.postorderTraversalByIter(root) == [3, 2, 1]

    assert solution.preorderTraversalByMonitor(root) == [1, 2, 3]
    assert solution.inorderTraversalByMonitor(root) == [1, 3, 2]
    assert solution.postorderTraversalByMonitor(root) == [3, 2, 1]

    print("All tests passed.")
